certgen/certgen.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package certgen generates (pre-)certificates and (pre-)certificate chains.
+package certgen
+
+import (
+	"crypto"
+	crand "crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"math"
+	"math/big"
+	"math/rand"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/certificate-transparency-go/x509"
+	"github.com/google/certificate-transparency-go/x509/pkix"
+	"github.com/google/monologue/interval"
+)
+
+var timeNowUTC = func() time.Time {
+	return time.Now().UTC()
+}
+
+// CertificateConfig contains details to be used to populate newly created leaf
+// certificates.
+type CertificateConfig struct {
+	// Required fields
+	//
+	// What these are set to, including the zero values if left unset, is what
+	// will appear in the leaf certificates.
+	SubjectCommonName         string
+	SubjectOrganization       string
+	SubjectOrganizationalUnit string
+	SubjectLocality           string
+	SubjectCountry            string
```

Comment:
Why not collapse all of these into a single `pkix.Name` field?

Reply:
Because it's more for the user of this package to have to do, to create a pkix.Name struct, and populate it correctly - including having to go and look up the pkix.Name struct etc.  There are also some additional fields in the pkix.Name struct that you wouldn't necessarily set, adding confusion for anyone who has to create one of those structs.  Having just the basic fields in the config seemed simpler for the user I think?  This feels like a bit of a six-of-one-but-half-a-dozen-of-another point tbh.